src/peft/utils/hotswap.py

```diff
@@ -13,23 +13,302 @@
+
+    return new_layer
+
+
+def _pad_lora_weights(model: torch.nn.Module, target_rank: int) -> None:
+    """
+    Pad LoRA weights in a model to a target rank while preserving the original behavior.
+
+    Args:
+      model (nn.Module): The model containing LoRA modules (with lora_A and lora_B).
+      target_rank (int): The target rank to pad to.
+    """
+
+    for module in model.modules():
+        # Decide which pad function to call based on module type
+        if isinstance(module, Linear):
+            pad_fn = _get_padded_linear
+        elif isinstance(module, Conv2d):
+            pad_fn = _get_padded_conv2d
+        else:
+            # Skip any other module types
+            continue
+
+        # Pad LoRA A
+        for adapter_name, lora_A_module in module.lora_A.items():
+            new_layer = pad_fn(lora_A_module, target_rank=target_rank, is_lora_A=True)
+            module.lora_A[adapter_name] = new_layer
+
+        # Pad LoRA B
+        for adapter_name, lora_B_module in module.lora_B.items():
+            new_layer = pad_fn(lora_B_module, target_rank=target_rank, is_lora_A=False)
+            module.lora_B[adapter_name] = new_layer
+
+
+def prepare_model_for_compiled_hotswap(
+    model: torch.nn.Module,
+    *,
+    target_rank: Optional[int] = None,
+    config: Optional[LoraConfig | dict[str, LoraConfig]] = None,
+) -> None:
+    """
+    Helper function that prepares the model so that it can later be compiled and then used with hot-swapping.
+
+    It is necessary to call this function on the model for hot-swapping to work if both of these are true:
+
+    - the different LoRA adapters have different ranks and/or different alpha values (i.e. scalings)
+    - you plan to torch.compile the model and want to avoid re-compilation
+
+    It is important to call this function *after* the first LoRA adapter has been loaded (i.e. the one that will be
+    swapped out) but *before* the model is compiled.
+
+    Even with this function, hot-swapping LoRA adapters that target different layers is still not supported.
+
+    Note: This function modifies the model in-place. If you want to restore the model to its initial state, you will
+    have to reload it.
+
+    Args:
+        model (`nn.Module`):
+            The model with the loaded adapter, before compilation.
+        target_rank (`int`, *optional*):
+            The target rank to pad the LoRA weights to. Should be the maximum rank among all LoRA adapters that will be
+            hot-swapped. If not specified, the target ranks will not be changed.
+        config (`LoraConfig` or `dict[str, LoraConfig]`, *optional*):
+            Optionally pass the `LoraConfig`s of the LoRA adapters. If passed, the rank in the configs will be updated
+            to `target_rank`.
+    """
+    is_compiled = hasattr(model, "_orig_mod")
+    if is_compiled:
+        raise ValueError("Call prepare_model_for_compiled_hotswap *before* compiling the model")
```

Comment:
Well, `_orig_mod` is present when we do `torch.compile()` but compilation can be also triggerd with `model.compile()`.

Reply:
We could check with [`torch.compiler.is_compiling()`](https://pytorch.org/docs/stable/generated/torch.compiler.is_compiling.html).

Reply:
Okay. Then maybe having a comment to mention the `.compile()` use case and that we don't know how to deal with that yet would be nice?

Reply:
Okay, I did find a way to check for `model.compile()`, namely the `_compiled_call_impl` attribute: https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/nn/modules/module.py#L2991

It is private, so I used `getattr` to ensure this doesn't break easily. I also tested this and it seems to work, although I'm not sure when users should use `model.compile` instead of `torch.compile(model)`.